ziti-tunnel/cmd/ziti-tunnel/subcmd/enroll.go

```diff
@@ -0,0 +1,156 @@
+	"github.com/spf13/cobra"
+	"io/ioutil"
+	"os"
+	"strings"
+)
+
+// global state used by all subcommands are located here for easy discovery
+var verbose bool
+var jwtpath, outpath, keyPath, certPath, idname, caOverride string
+
+const verboseDesc = "Enable verbose logging."
+const outpathDesc = "Output configuration file."
+const jwtpathDesc = "Enrollment token (JWT file). Required"
+const certDesc = "The certificate to present when establishing a connection."
+const idnameDesc = "Names the identity. Ignored if not 3rd party auto enrollment"
+
+const outFlag = "out"
+
+func init() {
+	enrollSubCmd.PersistentFlags().BoolVarP(&verbose, "verbose", "v", false, verboseDesc)
+	enrollSubCmd.Flags().StringVarP(&jwtpath, "jwt", "j", "", jwtpathDesc)
+	enrollSubCmd.Flags().StringVarP(&outpath, outFlag, "o", "", outpathDesc)
+	enrollSubCmd.Flags().StringVarP(&idname, "idname", "n", "", idnameDesc)
+	enrollSubCmd.Flags().StringVarP(&certPath, "cert", "c", "", certDesc)
+	enrollSubCmd.Flags().StringVarP(&caOverride, "ca", "", "", "Additional trusted certificates")
+
+	var keyDesc = ""
+	engines := certtools.ListEngines()
+	if len(engines) > 0 {
+		keyDesc = fmt.Sprintf("The key to use with the certifcate. Optionally specify the engine to use. supported engines: %v", engines)
+	} else {
+		keyDesc = fmt.Sprintf("The key to use with the certifcate.")
+	}
+
+	enrollSubCmd.Flags().StringVarP(&keyPath, "key", "k", "", keyDesc)
+
+	_ = enrollSubCmd.MarkFlagRequired("jwt")
+
+	root.AddCommand(enrollSubCmd)
+}
+
+var enrollSubCmd = &cobra.Command{
+	Use:   "enroll",
+	Short: "enroll an identity",
+	PersistentPreRun: func(cmd *cobra.Command, args []string) {
+		if verbose {
+			logrus.SetLevel(logrus.DebugLevel)
+		}
+	},
+	RunE: func(cmd *cobra.Command, args []string) error {
+		return processEnrollment()
+	},
+}
+
+func processEnrollment() error {
+	if strings.TrimSpace(outpath) == "" {
+		out, outErr := outPathFromJwt(jwtpath)
+		if outErr != nil {
+			return fmt.Errorf("could not set the output path: %s", outErr)
+		}
+		outpath = out
+	}
+
+	if jwtpath != "" {
+		if _, err := os.Stat(jwtpath); os.IsNotExist(err) {
+			return fmt.Errorf("the provided jwt file does not exist: %s", jwtpath)
+		}
+	}
+
+	if caOverride != "" {
+		if _, err := os.Stat(caOverride); os.IsNotExist(err) {
+			return fmt.Errorf("the provided ca file does not exist: %s", caOverride)
+		}
+	}
+
+	if strings.TrimSpace(outpath) == strings.TrimSpace(jwtpath) {
+		return fmt.Errorf("the output path must not be the same as the jwt path")
+	}
+
+	tokenStr, _ := ioutil.ReadFile(jwtpath)
+
+	pfxlog.Logger().Debugf("jwt to parse: %s", tokenStr)
+	tkn, _, err := enroll.ParseToken(string(tokenStr))
+
+	if err != nil {
+		return fmt.Errorf("failed to parse JWT: %s", err.Error())
+	}
+
+	flags := enroll.EnrollmentFlags{
+		CertFile:      certPath,
+		KeyFile:       keyPath,
+		Token:         tkn,
+		IDName:        idname,
+		AdditionalCAs: caOverride,
+	}
+	conf, err := enroll.Enroll(flags)
```

Comment:
Should we keep the stand-alone enroller too?  I've used it in the past, but have no problem using `ziti-tunnel` for an enrollment json I use somewhere else.  A little sad to have duplicate code, but updating `ziti-enroller` to call `ziti-tunnel` seems like a waste...

Reply:
yeah - i dunno. i added that same comment on the issue but not the pr:
> Add enroll capability to the ziti-tunnel and possibly deprecate the enroller - maybe

i would probably err on removing it with the 'we can put it back if we need it' idea

Reply:
`alias ziti-enroller=ziti-tunnel enroll $0`? :)

Reply:
I think ziti-tunnel could look at argv to see how it was called and go into enroller mode if args[0] is 'ziti-enroller`?